src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -3,91 +3,276 @@
+        PipeWriter pipeWriter,
+        AssetPath assetPath,
+        ReadOnlyMemory<Checksum> checksums,
+        SolutionAssetStorage.Scope scope,
+        ISerializerService serializer,
+        CancellationToken cancellationToken)
     {
-        public static async ValueTask WriteDataAsync(
-            Stream stream,
-            Dictionary<Checksum, object> assetMap,
-            ISerializerService serializer,
-            SolutionReplicationContext context,
-            Checksum solutionChecksum,
-            ReadOnlyMemory<Checksum> checksums,
-            CancellationToken cancellationToken)
+        // Create a channel to communicate between the searching and writing tasks.  This allows the searching task to
+        // find items, add them to the channel synchronously, and immediately continue searching for more items.
+        // Concurrently, the writing task can read from the channel and write the items to the pipe-writer.
+        var channel = Channel.CreateUnbounded<(Checksum checksum, object asset)>(new UnboundedChannelOptions()
         {
-            using var writer = new ObjectWriter(stream, leaveOpen: true, cancellationToken);
+            // We have a single task reading the data from the channel and writing it to the pipe.  This option
+            // allows the channel to operate in a more efficient manner knowing it won't have to sychronize data
+            // for multiple readers.
+            SingleReader = true,
+
+            // Currently we only have a single writer writing to the channel when we call FindAllAssetsAsync.
+            // However, we could change this in the future to allow the search to happen in parallel.
+            SingleWriter = true,
+        });
+
+        // When cancellation happens, attempt to close the channel.  That will unblock the task writing the assets
+        // to the pipe.
+        using var _ = cancellationToken.Register(() => channel.Writer.TryComplete(new OperationCanceledException(cancellationToken)));
+
+        // Keep track of how many checksums we found.  We must find all the checksums we were asked to find.
+        var foundChecksumCount = 0;
+
+        // Spin up a task to go search for all the requested checksums, adding results to the channel.
+        var findAssetsTask = FindAllAssetsAsync();
+
+        // Spin up a task to read from the channel and write out the assets to the pipe-writer.
+        var writeAssetsTask = WriteAllAssetsToPipeAsync();
 
-            // This information is not actually needed on the receiving end.  However, we still send it so that the
-            // receiver can assert that both sides are talking about the same solution snapshot and no weird invariant
-            // breaks have occurred.
-            solutionChecksum.WriteTo(writer);
+        // Wait for both the searching and writing tasks to finish.
+        await Task.WhenAll(findAssetsTask, writeAssetsTask).ConfigureAwait(false);
```

Comment:
>await Task.WhenAll(findAssetsTask, writeAssetsTask) [](http://example.com/codeflow?start=8&length=51)

nit: consider waiting separately to avoid the array alloc

Reply:
i prefer this form.  waiting separately also means exceptions from one might get lost.